### PR TITLE
treat interrupted compactions different than failures

### PR DESCRIPTION
**Pull request by lyonssp**
**What this PR does**:

This PR adds "interruption" as a new category of compaction outcome.  Currently, compaction starts and then it either fails or succeeds according to metrics.  In the event of a graceful shutdown, an ongoing compaction is recorded as a failure and `cortex_compactor_runs_failed_total` and `cortex_compactor_tenants_processing_failed` are incremented.  This makes it cumbersome to monitor for compaction failures.  After this PR, compactions interrupted by shutdown will not be recorded as failures but rather as "interruptions".

**Notes**:
Ideally, we could also verify in unit tests that `cortex_compactor_tenants_processing_failed` is not incremented, but this would require more complicated refactoring because `cortex_compactor_tenants_processing_*` metrics are reset at the end of compaction runs.

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/compactor/compactor.go [hunk 1]
```diff
@@ -670,6 +679,12 @@ func (c *Compactor) compactUsers(ctx context.Context) {
 		level.Info(c.logger).Log("msg", "starting compaction of user blocks", "user", userID)
 
 		if err = c.compactUserWithRetries(ctx, userID); err != nil {
+			if ctx.Err() != nil && ctx.Err() == context.Canceled {
```
Comment (alvinlin123) [inline pkg/compactor/compactor.go:682]: Wouldn't this mask the `err`?  Don't we still want to record down that there was an `err` in this case?

File: pkg/compactor/compactor.go [hunk 2]
```diff
@@ -640,7 +648,8 @@ func (c *Compactor) compactUsers(ctx context.Context) {
 	for _, userID := range users {
 		// Ensure the context has not been canceled (ie. compactor shutdown has been triggered).
 		if ctx.Err() != nil {
-			level.Info(c.logger).Log("msg", "interrupting compaction of user blocks", "err", err)
+			interrupted = true
+			level.Info(c.logger).Log("msg", "interrupting compaction of user blocks")
```
Comment (alvinlin123) [inline pkg/compactor/compactor.go:652]: While you are at it, can we enhance the log message to include value of `userId`? May come in handy for log diving later.

File: pkg/compactor/compactor.go [hunk 3]
```diff
@@ -640,7 +648,8 @@ func (c *Compactor) compactUsers(ctx context.Context) {
 	for _, userID := range users {
 		// Ensure the context has not been canceled (ie. compactor shutdown has been triggered).
 		if ctx.Err() != nil {
```
Comment (alvinlin123) [inline pkg/compactor/compactor.go:650]: Should we also add the `ctx.Err() == context.Canceled` check here? Or it's not needed?

Comment (lyonssp) [reply pkg/compactor/compactor.go:650]: It's not really clear whether we would benefit from a stricter check here, so this is a case where it's probably just best to not touch code we do not need to touch :)

File: pkg/compactor/compactor.go (hunk 2 above)
Comment (lyonssp) [reply pkg/compactor/compactor.go]: done, updated on both of the "interrupting compaction ..." log statements

File: pkg/compactor/compactor.go (hunk 1 above)
Comment (alvinlin123) [reply pkg/compactor/compactor.go:682]: I see. I think it's better to emit metrics such that we can detect the edge case you describe indeed happened (both failure and interrupted metrics were emitted); instead of guessing if it was the edge case when we see the interrupted metrics emitted.

Also, for line 650 above you mentioned:

> It's not really clear whether we would benefit from a stricter check

I a wondering what benefit do we have for stricter check here then?

Comment (lyonssp) [reply pkg/compactor/compactor.go:682]: > I think it's better to emit metrics such that we can detect the edge case you describe indeed happened (both failure and interrupted metrics were emitted); instead of guessing if it was the edge case when we see the interrupted metrics emitted.

Agreed that would be better but we don't have a good way to do that without refactoring.  Given the impact of hitting that edge case is a short delay in detecting a deterministic failure or the benefit of a retry in the case of transient errors, I don't see the value in refactoring.

If we emit both metrics, we are not fulfilling the core premise of this PR, which is to not signal compaction failure on restarts.

> I a wondering what benefit do we have for stricter check here then?

The stricter check here isn't just for added benefit, it fulfills the requirement.  A canceled context specifically indicates here that we are shutting down, whereas other context errors do not.

In the case of the other ctx.Err() check, I'm just saying it isn't clear what the original intent of that check is and I'd say it seems out of scope here to investigate and try to reason about whether we should update the existing code.

File: pkg/compactor/compactor.go [hunk 4]
```diff
@@ -607,6 +613,8 @@ func (c *Compactor) compactUsers(ctx context.Context) {
 		if succeeded && compactionErrorCount == 0 {
 			c.compactionRunsCompleted.Inc()
 			c.compactionRunsLastSuccess.SetToCurrentTime()
+		} else if interrupted {
+			c.compactionRunsInterrupted.Inc()
 		} else {
 			c.compactionRunsFailed.Inc()
```
Comment (yeya24) [inline pkg/compactor/compactor.go:619]: If there is failure before the interruption, we will only go L617 and miss `c.compactionRunsFailed` metric.
Probably increament the failure metric in a separate if block.

File: pkg/compactor/compactor.go [hunk 5]
```diff
@@ -607,6 +613,8 @@ func (c *Compactor) compactUsers(ctx context.Context) {
 		if succeeded && compactionErrorCount == 0 {
```
Comment (yeya24) [inline pkg/compactor/compactor.go:613]: If we introduce an additional `interrupted` var then seems `succeeded` is unncessary anymore.

File: pkg/compactor/compactor.go (hunk 1 above)
Comment (yeya24) [reply pkg/compactor/compactor.go:682]: https://github.com/prometheus/prometheus/pull/10772 What about using a similar way here to check the returned err instead of `ctx.Err()`?
If the edge case you meant is `ctx.Err() == ctx.Cancelled` whilte err is actually a compaction failure then this should catch it.

Comment (lyonssp) [reply pkg/compactor/compactor.go:682]: `errors.Is` will not work here because of the way that thanos wraps errors during compaction. 

Several of the intermediate errors that wrap failures during compaction like `compact.HaltError` and `errutil.NonNilMultiError` do not implement `.Unwrap()` and therefore do not work with `errors.Is` here.

TL;DR on how `errors.Is` works

`errors.Is(foo, bar)` calls `errors.Unwrap` on foo and its error chain until `errors.Unwrap` returns nil or `errors.Unwrap(x) == bar`.  If an error type does not implement `Unwrap()` then `errors.Unwrap()` returns nil.

File: pkg/compactor/compactor.go (hunk 4 above)
Comment (lyonssp) [reply pkg/compactor/compactor.go:619]: > If there is a failure before the interruption

In practice, this cannot currently happen. But I agree what you're saying makes sense in this local part of the code.  Will make the change.

File: pkg/compactor/compactor.go (hunk 1 above)
Comment (yeya24) [reply pkg/compactor/compactor.go:682]: What I am saying is to use the similar to check err rather than `ctx.Err()`. You can choose `Is` or `Cause` and see which one fits. Thanos also has `IsHaltError` for you to check whether it is halt error. If that's the only exception. And you can always implement `Unwrap` on the Thanos side, right?

Comment (lyonssp) [reply pkg/compactor/compactor.go:682]: > What I was saying is to use the similar to check err rather than ctx.Err(). You can choose Is or Cause and see which one fits

I think I responded directly to why this does not work in this case, let me know if I'm missing something.  All the problems with `Is` also exist with `Cause`, i.e. the underlying error types need to explicitly support these methods.

> Thanos also has IsHaltError for you to check whether it is halt error. If that's the only exception

Being able to check if the error is a `compact.HaltError` doesn't do anything for us because we still cannot unwrap HaltError to check whether `context.Canceled` is in the error chain.

> And you can always implement Unwrap on the Thanos side, right?

Sure, but I don't think Thanos updates should block this change, especially given this is already how we check for compactor interruption in the code.  There is precedent here.

Comment (yeya24) [reply pkg/compactor/compactor.go:682]: > I think I responded directly to why this does not work in this case, let me know if I'm missing something. All the problems with Is also exist with Cause, i.e. the underlying error types need to explicitly support these methods.

Anyway we are not checking err directly, we are checking `ctx.Err()` so anyway this doesn't solve the edge case, if this is the thing you and Alvin discussed. So I feel the best way is still to patch Thanos with the right method and check.

> given this is already how we check for compactor interruption in the code. There is precedent here.

If you mean the ctx check above I think it is more a check to handle closed ctx. For the wrapped error it is different.

File: pkg/compactor/compactor.go (hunk 4 above)
Comment (yeya24) [reply pkg/compactor/compactor.go:619]: Yes, that sounds good!

File: pkg/compactor/compactor.go (hunk 1 above)
Comment (yeya24) [reply pkg/compactor/compactor.go:682]: I don't want to block this. This is good as a first step. We can try to patch Thanos in a separate pr and update the code if necessary

**Review by yeya24 — APPROVED** [review]

File: pkg/compactor/compactor.go (hunk 4 above)
Comment (lyonssp) [reply pkg/compactor/compactor.go:619]: Cool, makes sense to me too.  Will write additional tests as well.

File: pkg/compactor/compactor.go (hunk 1 above)
Comment (lyonssp) [reply pkg/compactor/compactor.go:682]: I agree with you that updating Thanos to allow for better error checking here is ideal, for me it is just about whether or not we are blocking this.  I can log an action item to patch Thanos and come back to this.

Comment (yeya24) [reply pkg/compactor/compactor.go:682]: Yeah, let's not block it!

File: pkg/compactor/compactor.go (hunk 5 above)
Comment (lyonssp) [reply pkg/compactor/compactor.go]: We can actually do away with the error count as well and just have a `failed` boolean given we do not use the compactionErrorCount for anything than comparing with zero.  See latest updates.

**Review by alanprot — APPROVED** [review]
